ostd/src/mm/dma/dma_coherent.rs

```diff
@@ -1,186 +1,177 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use core::ops::Deref;
+use core::{fmt::Debug, mem::ManuallyDrop};
 
-use cfg_if::cfg_if;
-
-use super::{DmaError, check_and_insert_dma_mapping, remove_dma_mapping};
+use super::util::{alloc_kva, dma_remap, has_tdx, split_daddr, unmap_dma_remap};
+#[cfg(any(debug_assertions, all(target_arch = "x86_64", feature = "cvm_guest")))]
+use super::util::{alloc_unprotect_physical_range, dealloc_protect_physical_range};
 use crate::{
-    arch::iommu,
+    error::Error,
     mm::{
-        HasDaddr, HasPaddr, HasSize, Infallible, PAGE_SIZE, Paddr, USegment, VmReader, VmWriter,
-        dma::{Daddr, DmaType, dma_type},
+        Daddr, FrameAllocOptions, HasDaddr, HasPaddr, HasPaddrRange, HasSize, Infallible,
+        PAGE_SIZE, Paddr, Segment, Split, VmReader, VmWriter,
         io_util::{HasVmReaderWriter, VmReaderWriterIdentity},
-        kspace::{KERNEL_PAGE_TABLE, paddr_to_vaddr},
-        page_prop::CachePolicy,
+        kspace::kvirt_area::KVirtArea,
     },
 };
 
-cfg_if! {
-    if #[cfg(all(target_arch = "x86_64", feature = "cvm_guest"))] {
-        use crate::arch::tdx_guest;
-    }
-}
-
-/// A coherent (or consistent) DMA mapping,
-/// which guarantees that the device and the CPU can
-/// access the data in parallel.
-///
-/// The mapping will be destroyed automatically when
-/// the object is dropped.
+/// A DMA memory object with coherent cache.
 #[derive(Debug)]
 pub struct DmaCoherent {
-    segment: USegment,
-    start_daddr: Daddr,
+    inner: Inner,
+    map_daddr: Option<Daddr>,
     is_cache_coherent: bool,
 }
 
+#[derive(Debug)]
+enum Inner {
+    Segment(Segment<()>),
+    Kva(KVirtArea, Paddr),
+}
+
 impl DmaCoherent {
-    /// Creates a coherent DMA mapping backed by `segment`.
-    ///
-    /// The `is_cache_coherent` argument specifies whether
-    /// the target device that the DMA mapping is prepared for
-    /// can access the main memory in a CPU cache coherent way
-    /// or not.
```

Comment:
> access the memory with coherent access to the CPU cache

The new description is a bit of awkward: it includes two "access". It sounds as if the device explicitly accesses the CPU cache, but it is not, right? The memory controller or the system bus takes care of synchronizing the CPU cache with access memory. As far as the device is considered, it only "accesses" the memory. I don't think it is an improvement over the old description of "access the main memory in a CPU cache coherent way".